src/Definition.cpp

```diff
@@ -0,0 +1,210 @@
+    Schedule schedule;
+    ReductionDomain domain;
+    std::vector<Specialization> specializations;
+
+    DefinitionContents() : is_init(true) {}
+
+    void accept(IRVisitor *visitor) const {
+        for (Expr val : values) {
+            val.accept(visitor);
+        }
+        for (Expr arg : args) {
+            arg.accept(visitor);
+        }
+
+        schedule.accept(visitor);
+
+        if (domain.defined()) {
+            for (ReductionVariable rv : domain.domain()) {
+                rv.min.accept(visitor);
+                rv.extent.accept(visitor);
+            }
+            domain.predicate().accept(visitor);
+        }
+
+        for (const Specialization &s : specializations) {
+            if (s.condition.defined()) {
+                s.condition.accept(visitor);
+            }
+            s.definition.accept(visitor);
+        }
+    }
+
+    void mutate(IRMutator *mutator) {
+        for (size_t i = 0; i < values.size(); ++i) {
+            values[i] = mutator->mutate(values[i]);
+        }
+        for (size_t i = 0; i < args.size(); ++i) {
+            args[i] = mutator->mutate(args[i]);
+        }
+
+        schedule.mutate(mutator);
+
+        if (domain.defined()) {
+            domain.mutate(mutator);
+        }
+
+        for (Specialization &s : specializations) {
+            if (s.condition.defined()) {
+                s.condition = mutator->mutate(s.condition);
+            }
+            s.definition.mutate(mutator);
+        }
+    }
+};
+
+template<>
+EXPORT RefCount &ref_count<DefinitionContents>(const DefinitionContents *d) {
+    return d->ref_count;
+}
+
+template<>
+EXPORT void destroy<DefinitionContents>(const DefinitionContents *d) {
+    delete d;
+}
+
+Definition::Definition() : contents(new DefinitionContents) {}
+
+Definition::Definition(const IntrusivePtr<DefinitionContents> &ptr) : contents(ptr) {
+    internal_assert(ptr.defined())
+        << "Can't construct Function from undefined DefinitionContents ptr\n";
+}
+
+Definition::Definition(const std::vector<Expr> &args, const std::vector<Expr> &values,
+                       const ReductionDomain &rdom, bool is_init) : contents(new DefinitionContents) {
+    contents->is_init = is_init;
+    contents->values = values;
+    contents->args = args;
+    contents->domain = rdom;
+}
+
+// Return deep-copy of Definition
+Definition Definition::deep_copy(
+        std::map<IntrusivePtr<FunctionContents>, IntrusivePtr<FunctionContents>> &copied_map) const {
+    internal_assert(contents.defined()) << "Cannot deep-copy undefined Definition\n";
+
+    Definition copy;
+    copy.contents->is_init = contents->is_init;
+    copy.contents->values = contents->values;
+    copy.contents->args = contents->args;
+    copy.contents->schedule = contents->schedule.deep_copy(copied_map);
+
+    // Definition's domain is the same as the one pointed by its schedule.
+    internal_assert((contents->is_init == !contents->domain.defined()) || !contents->is_init)
```

Comment:
Also, I think the condition is slightly easier to understand if !contents->is_init is first, and then the second condition just assumes contents->is_init is true, so the condition becomes:

    internal_assert(!contents->is_init || !contents->domain.defined())